File: src/test/java/de/uni_mannheim/desq/mining/TraditionalMiningTest.java
package de.uni_mannheim.desq.mining;

import de.uni_mannheim.desq.dictionary.Dictionary;
import de.uni_mannheim.desq.io.DelPatternWriter;
import de.uni_mannheim.desq.io.SequenceReader;
import de.uni_mannheim.desq.util.DesqProperties;
import de.uni_mannheim.desq.util.TestUtils;
import org.apache.log4j.Logger;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;

import static org.assertj.core.api.Assertions.assertThat;

/** A test for traditional frequent sequence de.uni_mannheim.desq.old.mining. Datasets and parameters are set by implementing classes.
 *
 * Created by rgemulla on 18.07.2016.
 */
@RunWith(Parameterized.class)
public abstract class TraditionalMiningTest {
    private static final Logger logger = Logger.getLogger(TraditionalMiningTest.class);

    long sigma;
    int gamma, lambda;
    boolean generalize;
    String minerName;
    DesqProperties conf;

    TraditionalMiningTest(long sigma, int gamma, int lambda, boolean generalize,
                          String minerName, DesqProperties conf) {
        this.sigma = sigma;
        this.gamma = gamma;
        this.lambda = lambda;
        this.generalize = generalize;
        this.minerName = minerName;
        this.conf = conf;
    }

    /** The dictionary to use */
    public abstract Dictionary getDictionary() throws IOException;

    /** Whether counts and fids need to be computed from data */
    public abstract boolean computeStatisticsAndFids();

    /** Reader for data */
    public abstract SequenceReader getSequenceReader() throws IOException;

    public abstract String getGoldFileBaseName();

    public abstract String getTestDirectoryName();

    @Test
    public void test() throws IOException,
            InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
        String fileName = getGoldFileBaseName() + "-" + sigma + "-" + gamma + "-" + lambda + "-" + generalize + ".del";
        File actualFile = TestUtils.newTemporaryFile(
                TestUtils.getPackageResourcesPath(getClass()) + "/" + getTestDirectoryName() + "/" + minerName + "/" + fileName);
        mine(actualFile);
        try {
            File expectedFile = TestUtils.getPackageResource(getClass(), fileName);
            assertThat(actualFile).hasSameContentAs(expectedFile);
        } catch (NullPointerException e) {
            logger.error("Can't access expected data file for " + actualFile);
            throw e;
        }
    }

    public void mine(File outputDelFile)
            throws IOException, NoSuchMethodException, IllegalAccessException, InvocationTargetException,
                InstantiationException {
        // load the dictionary
        Dictionary dict = getDictionary();

        // update hierarchy
        if (computeStatisticsAndFids()) {
            SequenceReader dataReader = getSequenceReader();
            dataReader.setDictionary(dict);
            dict.clearCountsAndFids();
            dict.incCounts(dataReader);
            dataReader.close();
            dict.recomputeFids();
        }

        // Perform pattern de.uni_mannheim.desq.old.mining into del file
        SequenceReader dataReader = getSequenceReader();
        dataReader.setDictionary(dict);
        DesqMinerContext ctx = new DesqMinerContext();
        ctx.dict = dict;
        DelPatternWriter patternWriter = new DelPatternWriter(new FileOutputStream(outputDelFile), DelPatternWriter.TYPE.GID);
        patternWriter.setDictionary(dict);
        ctx.patternWriter = patternWriter;
        ctx.conf = conf;
        DesqMiner miner = DesqMiner.create(ctx);
        miner.addInputSequences(dataReader);
        dataReader.close();
        miner.mine();
        patternWriter.close();

        // sort del file
        TestUtils.sortDelPatternFile(outputDelFile);
    }
}

File: src/test/scala/de/uni_mannheim/desq/mining/spark/TraditionalMiningTest.scala
package de.uni_mannheim.desq.mining.spark

import java.io.{File, FileOutputStream}

import de.uni_mannheim.desq.io.DelPatternWriter
import de.uni_mannheim.desq.mining.WeightedSequence
import de.uni_mannheim.desq.util.{DesqProperties, TestUtils}
import org.apache.log4j.Logger
import org.apache.spark.SparkContext
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.scalatest.junit.AssertionsForJUnit


/** A test for traditional frequent sequence mining. Datasets and parameters are set by implementing classes.
 *
 * Created by rgemulla on 18.07.2016.
 */
@RunWith(classOf[Parameterized])
abstract class TraditionalMiningTest(sigma: Long, gamma: Int, lambda: Int, generalize: Boolean,
                                     minerName: String, conf: DesqProperties) extends AssertionsForJUnit {
    /** The data */
    def getDataset()(implicit sc: SparkContext): GenericDesqDataset[WeightedSequence]

    def goldFileBaseName: String

    def testDirectoryName: String

    @Test
    def test() {
        val fileName = goldFileBaseName + "-" + sigma + "-" + gamma + "-" + lambda + "-" + generalize + ".del"
        val actualFile = TestUtils.newTemporaryFile(
                TestUtils.getPackageResourcesPath(getClass) + "/" + testDirectoryName + "/" + minerName + "/" + fileName)
        mine(actualFile)
        try {
            val expectedFile = TestUtils.getPackageResource(classOf[de.uni_mannheim.desq.mining.TraditionalMiningTest], fileName) // use path of sequential results
            assertThat(actualFile).hasSameContentAs(expectedFile)
        } catch {
            case e: NullPointerException =>
                TraditionalMiningTest.logger.error("Can't access expected data file for " + actualFile)
                throw e
        }
    }

    def mine(outputDelFile: File) {
        implicit val sc = de.uni_mannheim.desq.util.spark.TestUtils.sc
        var data: GenericDesqDataset[WeightedSequence] = getDataset()

        // Perform pattern mining into del file
        val resultRDD = data.mine(conf)
        val result = resultRDD.sequences.collect()

        // write the data
        val patternWriter = new DelPatternWriter(new FileOutputStream(outputDelFile),
            DelPatternWriter.TYPE.GID)
        patternWriter.setDictionary(data.sequenceInterpreter.getDictionary)
        result.foreach(patternWriter.write)
        patternWriter.close()

        // sort del file
        TestUtils.sortDelPatternFile(outputDelFile)
    }
}

object TraditionalMiningTest {
    private val logger = Logger.getLogger(classOf[TraditionalMiningTest])
}
